sdk/core/azure-core/src/http/winhttp/win_http_request.hpp

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+// cspell:words PCCERT HCERTSTORE
+
+/**
+ * @file
+ * @brief #Azure::Core::Http::HttpTransport request support classes.
+ */
+
+#pragma once
+
+#include "azure/core/http/win_http_transport.hpp"
+#include "azure/core/url.hpp"
+#include <Windows.h>
+#include <memory>
+#include <mutex>
+#include <wil\resource.h>
+#include <winhttp.h>
+
+namespace Azure { namespace Core { namespace Http { namespace _detail {
+
+  class WinHttpRequest;
+  /**
+   * @brief An outstanding WinHTTP action. This object is used process asynchronous WinHTTP actions.
+   *
+   * The WinHttpRequest object has a WinHttpAction associated it to convert asynchronous WinHTTP
+   * operations to synchronous operations.
+   *
+   */
+  class WinHttpAction final {
+    friend class WinHttpRequest;
+
+    // Containing HTTP request, used during the status operation callback.
+    WinHttpRequest* m_httpRequest{};
+    std::mutex m_actionCompleteMutex;
+    //    std::condition_variable m_actionCompleted;
+    wil::unique_event m_actionCompleteEvent;
+    DWORD m_stowedError{};
+    DWORD_PTR m_stowedErrorInformation{};
+    DWORD m_bytesAvailable{};
+
+    /*
+     * Callback from WinHTTP called after the TLS certificates are received when the caller sets
+     * expected TLS root certificates.
+     */
+    static void CALLBACK StatusCallback(
+        HINTERNET hInternet,
+        DWORD_PTR dwContext,
+        DWORD dwInternetStatus,
+        LPVOID lpvStatusInformation,
+        DWORD dwStatusInformationLength) noexcept;
+
+    /*
+     * Callback from WinHTTP called after the TLS certificates are received when the caller sets
+     * expected TLS root certificates.
+     */
+    void OnHttpStatusOperation(
+        HINTERNET hInternet,
+        DWORD internetStatus,
+        LPVOID statusInformation,
+        DWORD statusInformationLength);
+
+  public:
+    /**
+     * @brief Create a new WinHttpAction object associated with a specific WinHttpRequest.
+     *
+     * @param request Http Request associated with the action.
+     */
+    WinHttpAction(WinHttpRequest* request)
+        : m_httpRequest(request), m_actionCompleteEvent(CreateEvent(nullptr, TRUE, FALSE, nullptr))
```

Comment:
maintainability nit: consider adding a comment describing why we are passing these options to `CreateEvent`